README.md
# Argument Handling for PHP CLI Scripts

This project aims to deliver a easy to use and free as in freedom php command component.

The build status of the current master branch is tracked by Travis CI:
[![Build Status](https://travis-ci.org/bazzline/php_component_cli_arguments.png?branch=master)](http://travis-ci.org/bazzline/php_component_cli_arguments)
[![Latest stable](https://img.shields.io/packagist/v/net_bazzline/php_component_cli_arguments.svg)](https://packagist.org/packages/net_bazzline/php_component_cli_arguments)

The scrutinizer status are:
[![code quality](https://scrutinizer-ci.com/g/bazzline/php_component_cli_arguments/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/bazzline/php_component_cli_arguments/) | [![build status](https://scrutinizer-ci.com/g/bazzline/php_component_cli_arguments/badges/build.png?b=master)](https://scrutinizer-ci.com/g/bazzline/php_component_cli_arguments/)

The versioneye status is:
[![Dependency Status](https://www.versioneye.com/user/projects/553941560b24225ef6000002/badge.svg?style=flat)](https://www.versioneye.com/user/projects/553941560b24225ef6000002)

Take a look on [openhub.net](https://www.openhub.net/p/php_component_cli_arguments).

The current change log can be found [here](https://github.com/bazzline/php_component_cli_arguments/blob/master/CHANGELOG.md).

# Install

## By Hand

```bash
mkdir -p vendor/net_bazzline/php_component_cli_arguments
cd vendor/net_bazzline/php_component_cli_arguments
git clone https://github.com/bazzline/php_component_cli_arguments .
```

## With [Packagist](https://packagist.org/packages/net_bazzline/php_component_cli_arguments)

```bash
composer require net_bazzline/php_component_cli_arguments:dev-master
```

# Benefits

* easy up handling of following kinds of arguments
    * flags (command -f|--force)
    * lists (command --foobar=foo | command -f=foo)
    * values values (command <value>)

# Example

Simple call [run.php](https://github.com/bazzline/php_component_cli_arguments/blob/master/Example/run.php) with tons of arguments like illustrated below.

```
php run.php --foo bar --foobar=foo --foobar="bar" -f="foo" -f=bar -b foobar foo -flag
```

Generates the following output.
```
arguments provided:
    --foo
    bar
    --foobar=foo
    --foobar=bar
    -f=foo
    -f=bar
    -b
    foobar
    foo
    -flag
flags provided:
    foo
    b
    f
    l
    a
    g
lists provided:
    foobar
        foo
        bar
    f
        foo
        bar
values provided:
    bar
    foobar
    foo
```

# Terms

All arguments are grouped into one of three types, flags, lists or values.
[Argument](http://en.wikipedia.org/wiki/Argument) or [parameter](http://en.wikipedia.org/wiki/Parameter)? I could not spot a major difference. When I think about parameters, my mind slipps into the domain of methods or functions, thats why I have decided to call them arguments. Furthermore, [php.net](http://php.net/manual/en/function.getopt.php) calls the "argument list" also :-).

## Flag

A flag is an argument that changes the behaviour of an command. It acts as a trigger so you can turn things on or off (best example in the world "-h|--help")

The position in a commandcall for a flag is not important, only the existence.

Valid flags are:

* -f
* --flag
* -flag (shortcut for -f -l -a -g)

## List

A list is an argument that contains multiple values per name.

This call
```
php example.php --my_list="argument one" --my_list="argument two"
```
would result into a list with the name "my_list" and two arguments, "argument one" and "argument two".

Lists are the most complex arguments. Like for flags, the position in a commandcall for a list usage is not important.

Valid Lists are:

* -l=value
* -l="val ue"
* --list=value
* --list="val ue"

## Value

Values are straight forward arguments. You simple pass them to your command. Instead of a flag or a list, the position is important.

Valid values are:

* value
* "val ue"

```
php example.php "value one" "value two"
```

First value has the content "value one", second value has the content "value two".

```
php example.php "value two" "value one"
```

First value has the content "value two", second value has the content "value one".

## Short Name and Long Name Notation

Flag and list arguments supporting short name ("-f") and long name ("--foo") notation.
A short name is indicated by a single "-" while a long name is indicated by a double "-".
The handling and the support of them is domain specific (and also a matter of tast). To merge the usage and the content for lists is not part of this component.

# Why no Validation?

Validation is a complex topic. That's why I decided to not put it into the domain of this component.  
It would complicate the code itself. I would have created a universal validation interface that would slow down the usage of this component. Furthermore, you would have to learn a validation expression language or would have need to write code that fits my validation interface but not your "way of coding".

At the end, what is validation all about?
* check if a argument (flag, list, value) is passed or not
* if it is passed validate the value or if it is allowed under that circumstance (if it is right to use flag "-f" while also flag "-b" is passed etc.)
* if it is not passed but was mandatory, create a specific message or throw an exception (and the same for optional arguments)

To sum it up, validation is domain specific for the validation itself and the error handling. That's why I have decided to not support it deeply. The component supports your validation implementation with the methods "hasLists()", "hasList($name)" etc.

Since I won't write "never say never", if you have a smart idea or way to easy up validation, I'm open for an question or a pull request.

# What about Optional Arguments?

Optional arguments underlying the same problems as validation. It is not that easy to implement in an elegant way. It is very special/domain specific (e.g. an argument is optional if flag "--xyz" is used, otherwise mandatory).
Your code has to take care if an argument is passed or not anyways. Using the available "has..."-methods should be sufficient and generic enough.

# API

[API](http://www.bazzline.net/988fc4501e48d8aed30583f58a97f0d4c14ab2d3/index.html) is available at [bazzline.net](http://www.bazzline.net).

# Other Great Components

* [search on packagist](https://packagist.org/search/?search_query%5Bquery%5D=getopt)
* [yeriomin/getopt](https://github.com/yeriomin/getopt)
* [ulrichsg/getopt-php](https://github.com/ulrichsg/getopt-php)
* [stuartherbert/CommandLineLib](https://github.com/stuartherbert/CommandLineLib/)
* [auraphp/Aura.Cli](https://github.com/auraphp/Aura.Cli)
* [hoaproject/Console](https://github.com/hoaproject/Console)
* [deweller/php-cliopts](https://github.com/deweller/php-cliopts)

# Final Words

Star it if you like it :-). Add issues if you need it. Pull patches if you enjoy it. Write a blog entry if you use it :-D.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Open]

### To Add

* add *hasFlags* to easy up validation if long or short flag is set (e.g. '-v|--verbose')

### To Change

* cover Parser with unit tests

## [Unreleased]

### Added

* added nightly to the travis.yml

### Changeed

## [2.1.0](https://github.com/bazzline/php_component_cli_argument/tree/2.1.0) - released at 2022-11-23

### Changeed

* Raised minimum php version requirement from `7.0` to `8.0`
* Updated development requirement from phpunit 6.0 to 8.0
* Migrated phpunit xml
* Migrated code base to php 8.0

## [2.0.0](https://github.com/bazzline/php_component_cli_argument/tree/2.0.0) - released at 2018-01-12

### Changed

* dropped support for all PHP Versions below 7.0

## [1.4.1](https://github.com/bazzline/php_component_cli_argument/tree/1.4.1) - released at 2017-05-18

### Changed

* converted history into changelog

## [1.4.0](https://github.com/bazzline/php_component_cli_argument/tree/1.4.0) - released at 2017-01-29

### Changed

* updated minimum requirements to php 5.6

## [1.3.2](https://github.com/bazzline/php_component_cli_argument/tree/1.3.2) - released at 2016-08-07

### Changed

* updated to phpunit 5.5

## [1.3.1](https://github.com/bazzline/php_component_cli_argument/tree/1.3.1) - released at 2015-06-05

### Added

* added php 7.0 to testing environment

### Changed

* enhanced phpunit compatibility by using "~4.8||~5.4"
* moved to psr-4 autoloading
* removed php 5.3.3 testing environment
* updated to phpunit 5.4

## [1.3.0](https://github.com/bazzline/php_component_cli_argument/tree/1.3.0) - released at 2015-12-01

### Added

* added support for single argument like "a" or "-"

## [1.2.0](https://github.com/bazzline/php_component_cli_argument/tree/1.2.0) - released at 2015-11-28

### Added

* added *convertToArray()'
* added *convertToString()'
* added *getNumberOf[Arguments|Flags|Lists|Values]

### Changed

* removed *generate_api*
* replaced "setArguments" with "parseArguments" like [nette](https://github.com/nette/command-line/blob/master/src/CommandLine/Parser.php) is doing it
* updated dependency handling by being less restrictive (added lower limit for phpunit)

## [1.1.2](https://github.com/bazzline/php_component_cli_argument/tree/1.1.2) - released at 2015-11-07

### Changed

* updated dependencies

## [1.1.1](https://github.com/bazzline/php_component_cli_argument/tree/1.1.1) - released at 2015-08-18

### Changed

* updated dependencies

## [1.1.0](https://github.com/bazzline/php_component_cli_argument/tree/1.1.0) - released at 2015-07-02

### Changed

* added second argument to `Arguments::__construct()` and `Arguments::setArguments` called "removeFirstArgument"

## [1.0.2](https://github.com/bazzline/php_component_cli_argument/tree/1.0.2) - released at 2015-07-02

### Changed

* updated dependencies

## [1.0.1](https://github.com/bazzline/php_component_cli_argument/tree/1.0.1) - released at 2015-05-22

### Changed

* updated dependencies

## [1.0.0](https://github.com/bazzline/php_component_cli_argument/tree/1.0.0) - released at 2015-04-23

### Added

* initial release


test/ArgumentsTest.php
<?php
/**
 * @author stev leibelt <artodeto@bazzline.net>
 * @since 2015-04-18 
 */

namespace Test\Net\Bazzline\Component\Cli\Arguments;

use Net\Bazzline\Component\Cli\Arguments\Arguments;
use PHPUnit\Framework\TestCase;

class ArgumentsTest extends TestCase
{
    public function testWithNoArgv()
    {
        $arguments = $this->createArguments();

        $this->assertFalse($arguments->hasArguments());
        $this->assertFalse($arguments->hasLists());
        $this->assertFalse($arguments->hasFlags());
        $this->assertFalse($arguments->hasValues());
    }

    /**
     * @return array
     */
    public function withArgumentsProvider()
    {
        return [
            'empty argv' => [
                'argv'      => [],
                'arguments' => [],
                'flags'     => [],
                'lists'     => [],
                'values'    => [],
            ],
            'only file name argument' => [
                'argv'      => [
                    __FILE__
                ],
                'arguments' => [],
                'flags'     => [],
                'lists'     => [],
                'values'    => []
            ],
            'one value with one character' => [
                'argv'      => [
                    __FILE__,
                    '-'
                ],
                'arguments' => [
                    '-'
                ],
                'flags'     => [],
                'lists'     => [],
                'values'    => [
                    '-'
                ]
            ],
            'one value' => [
                'argv'      => [
                    __FILE__,
                    'foo'
                ],
                'arguments' => [
                    'foo'
                ],
                'flags'     => [],
                'lists'     => [],
                'values'    => [
                    'foo'
                ]
            ],
            'one short flag' => [
                'argv'      => [
                    __FILE__,
                    '-f'
                ],
                'arguments' => [
                    '-f'
                ],
                'flags'     => [
                    'f'
                ],
                'lists'     => [],
                'values'    => []
            ],
            'one long flag' => [
                'argv'      => [
                    __FILE__,
                    '--foobar'
                ],
                'arguments' => [
                    '--foobar'
                ],
                'flags'     => [
                    'foobar'
                ],
                'lists'     => [],
                'values'    => []
            ],
            'one short list without quotation mark' => [
                'argv'      => [
                    __FILE__,
                    '-f=oo'
                ],
                'arguments' => [
                    '-f=oo'
                ],
                'flags'     => [],
                'lists'     => [
                    'f' => [
                        'oo'
                    ]
                ],
                'values'    => []
            ],
            'one short list with quotation mark' => [
                'argv'      => [
                    __FILE__,
                    '-f="oo"'
                ],
                'arguments' => [
                    '-f="oo"'
                ],
                'flags'  => [],
                'lists'     => [
                    'f' => [
                        'oo'
                    ]
                ],
                'values'    => []
            ],
            'one long list without quotation mark' => [
                'argv'      => [
                    __FILE__,
                    '--foobar=baz'
                ],
                'arguments' => [
                    '--foobar=baz'
                ],
                'flags'  => [],
                'lists'     => [
                    'foobar' => [
                        'baz'
                    ]
                ],
                'values'    => []
            ],
            'one long list with quotation mark' => [
                'argv'      => [
                    __FILE__,
                    '--foobar="baz"'
                ],
                'arguments' => [
                    '--foobar="baz"'
                ],
                'flags'  => [],
                'lists'     => [
                    'foobar' => [
                        'baz'
                    ]
                ],
                'values'    => []
            ],
            'complex example' => [
                'argv'      => [
                    __FILE__,
                    '--foobar="foo"',
                    '--foobar=bar',
                    'foobar',
                    '-f=foo',
                    '-f="bar"',
                    '-b',
                    'foo',
                    '-z',
                    '-flag'
                ],
                'arguments' => [
                    '--foobar="foo"',
                    '--foobar=bar',
                    'foobar',
                    '-f=foo',
                    '-f="bar"',
                    '-b',
                    'foo',
                    '-z',
                    '-flag'
                ],
                'flags'  => [
                    'b',
                    'z',
                    'f',
                    'l',
                    'a',
                    'g'
                ],
                'lists'     => [
                    'foobar'    => [
                        'foo',
                        'bar'
                    ],
                    'f'         => [
                        'foo',
                        'bar'
                    ]
                ],
                'values'    => [
                    'foobar',
                    'foo'
                ]
            ]
        ];
    }

    /**
     * @dataProvider withArgumentsProvider
     * @param array $argv
     * @param array $expectedArguments
     * @param array $expectedFlags
     * @param array $expectedLists
     * @param array $expectedValues
     */
    public function testWithArguments(
        array $argv,
        array $expectedArguments,
        array $expectedFlags,
        array $expectedLists,
        array $expectedValues
    ) {
        $arguments = $this->createArguments($argv);

        $this->assertEquals((!empty($expectedArguments)), $arguments->hasArguments());
        $this->assertEquals((!empty($expectedFlags)), $arguments->hasFlags());
        $this->assertEquals((!empty($expectedLists)), $arguments->hasLists());
        $this->assertEquals((!empty($expectedValues)), $arguments->hasValues());

        $this->assertEquals((count($expectedArguments)), $arguments->getNumberOfArguments());
        $this->assertEquals((count($expectedFlags)), $arguments->getNumberOfFlags());
        $this->assertEquals((count($expectedLists)), $arguments->getNumberOfLists());
        $this->assertEquals((count($expectedValues)), $arguments->getNumberOfValues());

        $this->assertEquals($expectedArguments, $arguments->getArguments());
        $this->assertEquals($expectedFlags, $arguments->getFlags());
        $this->assertEquals($expectedLists, $arguments->getLists());
        $this->assertEquals($expectedValues, $arguments->getValues());

        foreach ($expectedFlags as $name) {
            $this->assertTrue($arguments->hasFlag($name));
        }

        foreach ($expectedLists as $name => $values) {
            $this->assertTrue($arguments->hasList($name));
            $this->assertEquals($values, $arguments->getList($name));
        }
    }

    public function testWithArgumentsAndNotRemovingFirstArgument()
    {
        $argv       = [
            'foo',
            'bar'
        ];
        $arguments  = $this->createArguments($argv, false);

        $this->assertEquals($argv, $arguments->getArguments());
        $this->assertEquals($argv, $arguments->getValues());
        $this->assertEmpty($arguments->getFlags());
        $this->assertEmpty($arguments->getLists());
    }

    /**
     * @param null|array $argv
     * @param boolean $removeFirstArgument
     * @return Arguments
     */
    private function createArguments($argv = null, $removeFirstArgument = true)
    {
        return new Arguments($argv, $removeFirstArgument);
    }
}


test/Test/Net/Bazzline/Component/Cli/Arguments/ArgumentsTest.php
<?php
/**
 * @author stev leibelt <artodeto@bazzline.net>
 * @since 2015-04-18 
 */

namespace Test\Net\Bazzline\Component\Cli\Arguments;

use Net\Bazzline\Component\Cli\Arguments\Arguments;
use PHPUnit_Framework_TestCase;

class ArgumentsTest extends PHPUnit_Framework_TestCase
{
    public function testWithNoArgv()
    {
        $arguments = $this->createArguments();

        $this->assertFalse($arguments->hasArguments());
        $this->assertFalse($arguments->hasLists());
        $this->assertFalse($arguments->hasFlags());
        $this->assertFalse($arguments->hasValues());
    }

    /**
     * @return array
     */
    public function testWithArgumentsProvider()
    {
        return array(
            'empty argv' => array(
                'argv'      => array(),
                'arguments' => array(),
                'flags'     => array(),
                'lists'     => array(),
                'values'    => array()
            ),
            'only file name argument' => array(
                'argv'      => array(
                    __FILE__
                ),
                'arguments' => array(),
                'flags'     => array(),
                'lists'     => array(),
                'values'    => array()
            ),
            'one value with one character' => array(
                'argv'      => array(
                    __FILE__,
                    '-'
                ),
                'arguments' => array(
                    '-'
                ),
                'flags'     => array(),
                'lists'     => array(),
                'values'    => array(
                    '-'
                )
            ),
            'one value' => array(
                'argv'      => array(
                    __FILE__,
                    'foo'
                ),
                'arguments' => array(
                    'foo'
                ),
                'flags'     => array(),
                'lists'     => array(),
                'values'    => array(
                    'foo'
                )
            ),
            'one short flag' => array(
                'argv'      => array(
                    __FILE__,
                    '-f'
                ),
                'arguments' => array(
                    '-f'
                ),
                'flags'     => array(
                    'f'
                ),
                'lists'     => array(),
                'values'    => array()
            ),
            'one long flag' => array(
                'argv'      => array(
                    __FILE__,
                    '--foobar'
                ),
                'arguments' => array(
                    '--foobar'
                ),
                'flags'     => array(
                    'foobar'
                ),
                'lists'     => array(),
                'values'    => array()
            ),
            'one short list without quotation mark' => array(
                'argv'      => array(
                    __FILE__,
                    '-f=oo'
                ),
                'arguments' => array(
                    '-f=oo'
                ),
                'flags'     => array(),
                'lists'     => array(
                    'f' => array(
                        'oo'
                    )
                ),
                'values'    => array()
            ),
            'one short list with quotation mark' => array(
                'argv'      => array(
                    __FILE__,
                    '-f="oo"'
                ),
                'arguments' => array(
                    '-f="oo"'
                ),
                'flags'  => array(),
                'lists'     => array(
                    'f' => array(
                        'oo'
                    )
                ),
                'values'    => array()
            ),
            'one long list without quotation mark' => array(
                'argv'      => array(
                    __FILE__,
                    '--foobar=baz'
                ),
                'arguments' => array(
                    '--foobar=baz'
                ),
                'flags'  => array(),
                'lists'     => array(
                    'foobar' => array(
                        'baz'
                    )
                ),
                'values'    => array()
            ),
            'one long list with quotation mark' => array(
                'argv'      => array(
                    __FILE__,
                    '--foobar="baz"'
                ),
                'arguments' => array(
                    '--foobar="baz"'
                ),
                'flags'  => array(),
                'lists'     => array(
                    'foobar' => array(
                        'baz'
                    )
                ),
                'values'    => array()
            ),
            'complex example' => array(
                'argv'      => array(
                    __FILE__,
                    '--foobar="foo"',
                    '--foobar=bar',
                    'foobar',
                    '-f=foo',
                    '-f="bar"',
                    '-b',
                    'foo',
                    '-z',
                    '-flag'
                ),
                'arguments' => array(
                    '--foobar="foo"',
                    '--foobar=bar',
                    'foobar',
                    '-f=foo',
                    '-f="bar"',
                    '-b',
                    'foo',
                    '-z',
                    '-flag'
                ),
                'flags'  => array(
                    'b',
                    'z',
                    'f',
                    'l',
                    'a',
                    'g'
                ),
                'lists'     => array(
                    'foobar'    => array(
                        'foo',
                        'bar'
                    ),
                    'f'         => array(
                        'foo',
                        'bar'
                    )
                ),
                'values'    => array(
                    'foobar',
                    'foo'
                )
            )
        );
    }

    /**
     * @dataProvider testWithArgumentsProvider
     * @param array $argv
     * @param array $expectedArguments
     * @param array $expectedFlags
     * @param array $expectedLists
     * @param array $expectedValues
     */
    public function testWithArguments(
        array $argv,
        array $expectedArguments,
        array $expectedFlags,
        array $expectedLists,
        array $expectedValues
    )
    {
        $arguments = $this->createArguments($argv);

        $this->assertEquals((!empty($expectedArguments)), $arguments->hasArguments());
        $this->assertEquals((!empty($expectedFlags)), $arguments->hasFlags());
        $this->assertEquals((!empty($expectedLists)), $arguments->hasLists());
        $this->assertEquals((!empty($expectedValues)), $arguments->hasValues());

        $this->assertEquals((count($expectedArguments)), $arguments->getNumberOfArguments());
        $this->assertEquals((count($expectedFlags)), $arguments->getNumberOfFlags());
        $this->assertEquals((count($expectedLists)), $arguments->getNumberOfLists());
        $this->assertEquals((count($expectedValues)), $arguments->getNumberOfValues());

        $this->assertEquals($expectedArguments, $arguments->getArguments());
        $this->assertEquals($expectedFlags, $arguments->getFlags());
        $this->assertEquals($expectedLists, $arguments->getLists());
        $this->assertEquals($expectedValues, $arguments->getValues());

        foreach ($expectedFlags as $name) {
            $this->assertTrue($arguments->hasFlag($name));
        }

        foreach ($expectedLists as $name => $values) {
            $this->assertTrue($arguments->hasList($name));
            $this->assertEquals($values, $arguments->getList($name));
        }
    }

    public function testWithArgumentsAndNotRemovingFirstArgument()
    {
        $argv       = array('foo', 'bar');
        $arguments  = $this->createArguments($argv, false);

        $this->assertEquals($argv, $arguments->getArguments());
        $this->assertEquals($argv, $arguments->getValues());
        $this->assertEmpty($arguments->getFlags());
        $this->assertEmpty($arguments->getLists());
    }

    /**
     * @param null|array $argv
     * @param boolean $removeFirstArgument
     * @return Arguments
     */
    private function createArguments($argv = null, $removeFirstArgument = true)
    {
        return new Arguments($argv, $removeFirstArgument);
    }
}


Example/run.php
<?php
/**
 * @author: stev leibelt <artodeto@bazzline.net>
 * @since: 2015-04-16
 */

use Net\Bazzline\Component\Cli\Arguments\Arguments;

require_once __DIR__ . '/../vendor/autoload.php';

$arguments = new Arguments($argv);

if ($arguments->hasArguments()) {
    echo $arguments->getNumberOfArguments() . ' arguments provided:' . PHP_EOL;
    foreach ($arguments->getArguments() as $argument) {
        echo '    ' . $argument . PHP_EOL;
    }
} else {
    echo 'no values provided' . PHP_EOL;
}

if ($arguments->hasLists()) {
    echo $arguments->getNumberOfLists() . ' lists provided:' . PHP_EOL;
    foreach ($arguments->getLists() as $name => $values) {
        echo '    ' . $name . PHP_EOL;
        foreach ($values as $value) {
            echo '        ' . $value . PHP_EOL;
        }
    }
} else {
    echo 'no values provided' . PHP_EOL;
}

if ($arguments->hasFlags()) {
    echo $arguments->getNumberOfFlags() . ' flags provided:' . PHP_EOL;
    foreach ($arguments->getFlags() as $flag) {
        echo '    ' . $flag . PHP_EOL;
    }
} else {
    echo 'no flags provided' . PHP_EOL;
}

if ($arguments->hasValues()) {
    echo $arguments->getNumberOfValues() . ' values provided:' . PHP_EOL;
    foreach ($arguments->getValues() as $value) {
        echo '    ' . $value . PHP_EOL;
    }
} else {
    echo 'no values provided' . PHP_EOL;
}


source/Parser.php
<?php
/**
 * @author stev leibelt <artodeto@bazzline.net>
 * @since 2015-11-26
 */
namespace Net\Bazzline\Component\Cli\Arguments;

class Parser
{
    private array $flags;
    private array $lists;
    private array $values;

    public function __construct()
    {
        $this->initiate();
    }

    public function parse(array $arguments): void
    {
        $this->initiate();

        foreach ($arguments as $argument) {
            $this->addToFittingCollection($argument);
        }
    }

    public function getFlags(): array
    {
        return $this->flags;
    }

    public function getLists(): array
    {
        return $this->lists;
    }

    public function getValues(): array
    {
        return $this->values;
    }

    private function addToList(string $name, mixed $value): void
    {
        $value = trim($value, '"'); //remove >"< if exists

        $collection = $this->lists[$name] ?? [];

        $collection[]       = $value;
        $this->lists[$name] = $collection;
    }

    private function contains(string $string, string $search): bool
    {
        if (strlen($search) === 0) {
            $contains = false;
        } else {
            $contains = !(!str_contains($string, $search));
        }

        return $contains;
    }

    private function handleLongNameListOrFlag(string $argument): void
    {
        if ($this->contains($argument, '=')) {
            $position   = strpos($argument, '=');
            $name       = substr($argument, 0, $position);
            $value      = substr($argument, ($position + 1));
            $this->addToList($name, $value);
        } else {
            $this->flags[] = $argument;
        }
    }

    private function handleShortNameListOrFlag(string $argument): void
    {
        $containsEqualCharacter             = ($this->contains($argument, '='));
        $equalCharacterIsOnSecondPosition   = (strpos($argument, '=') === 1);
        $isShortNameList                    = ($containsEqualCharacter
            && $equalCharacterIsOnSecondPosition);

        if ($isShortNameList) {
            $name   = substr($argument, 0, 1);
            $value  = substr($argument, 2);
            $this->addToList($name, $value);
        } else if (!$containsEqualCharacter) {
            $length = strlen($argument);
            $iterator = 0;
            while ($iterator < $length) {
                $this->flags[] = $argument[$iterator];
                ++$iterator;
            }
        }
    }

    private function initiate(): void
    {
        $this->flags    = [];
        $this->lists    = [];
        $this->values   = [];
    }

    private function addToFittingCollection(string $argument): void
    {
        if ($this->hasLengthOf($argument, 1)) {
            $this->values[] = $argument;
        } else if ($this->startsWith($argument, '--')) {
            $argument = substr($argument, 2);
            $this->handleLongNameListOrFlag($argument);
        } else if ($this->startsWith($argument, '-')) {
            $argument = substr($argument, 1);
            $this->handleShortNameListOrFlag($argument);
        } else {
            $this->values[] = $argument;
        }
    }

    private function startsWith(string $string, string $start): bool
    {
        return (strncmp($string, $start, strlen($start)) === 0);
    }

    private function hasLengthOf(string $string, int $expectedLength): bool
    {
        $length = strlen($string);

        return ($length == $expectedLength);
    }
}


example.php
<?php
/**
 * @author: stev leibelt <artodeto@bazzline.net>
 * @since: 2015-04-16
 */

require_once __DIR__ . '/Arguments.php';

$arguments = new Arguments($argv);

if ($arguments->hasArguments()) {
    echo 'arguments provided:' . PHP_EOL;
    foreach ($arguments->getArguments() as $argument) {
        echo '    ' . $argument . PHP_EOL;
    }
} else {
    echo 'no values provided' . PHP_EOL;
}

if ($arguments->hasLists()) {
    echo 'lists provided:' . PHP_EOL;
    foreach ($arguments->getLists() as $name => $values) {
        echo '    ' . $name . PHP_EOL;
        foreach ($values as $value) {
            echo '        ' . $value . PHP_EOL;
        }
    }
} else {
    echo 'no values provided' . PHP_EOL;
}

if ($arguments->hasFlags()) {
    echo 'flags provided:' . PHP_EOL;
    foreach ($arguments->getFlags() as $flag) {
        echo '    ' . $flag . PHP_EOL;
    }
} else {
    echo 'no flags provided' . PHP_EOL;
}

if ($arguments->hasValues()) {
    echo 'values provided:' . PHP_EOL;
    foreach ($arguments->getValues() as $value) {
        echo '    ' . $value . PHP_EOL;
    }
} else {
    echo 'no values provided' . PHP_EOL;
}